datasets/cartesian/data_generation.py

```diff
@@ -0,0 +1,277 @@
+  test_easy_examples, it_dec_test_easy_examples = create_examples(testsize, 
+                                                                  trainmindigits, 
+                                                                  trainmaxdigits)
+  test_hard_examples, it_dec_test_hard_examples = create_examples(testsize, 
+                                                                  testmindigits, 
+                                                                  testmaxdigits)
+  return (train_examples, it_dec_train_examples, 
+          test_easy_examples, it_dec_test_easy_examples,
+          test_hard_examples, it_dec_test_hard_examples)
+
+
+def main():
+
+  def generate_original_data_files(examples_in, examples_out, filepath_src, 
+                                  filepath_tgt):
+    out_file_src = open(filepath_src, "w")
+    out_file_tgt = open(filepath_tgt, "w")
+    for in_line, out_line in zip(examples_in, examples_out):
+      out_file_src.write(" ".join(in_line) + "\n")
+      out_file_tgt.write(" ".join(out_line) + "\n")
+    out_file_src.close()
+    out_file_tgt.close()
+
+
+  def generate_it_dec_data_files(examples_in, examples_out, filepath_src, 
+                                  filepath_tgt):
+    out_file_src = open(filepath_src, "w")
+    out_file_tgt = open(filepath_tgt, "w")
+    for in_line, out_line in zip(examples_in, examples_out):
+      for l1, l2 in zip(in_line, out_line):
+        out_file_src.write(" ".join(l1) + "\n")
+        out_file_tgt.write(" ".join(l2) + "\n")
+    out_file_src.close()
+    out_file_tgt.close()
+
+
+  examples = create_cartesian_dataset(trainsize = 200000, testsize = 1024, 
+                                      trainmindigits = 1, trainmaxdigits = 6, 
+                                      testmindigits = 7, testmaxdigits = 8, 
+                                      repeat_digits = False, reverse = False, 
+                                      row = row, copy_output = copy, 
+                                      short_input = short_input)
+  # Generating (normal) training data.
+  generate_original_data_files(examples[0][0], examples[0][1], data_path + "train.src",
+                               data_path + "train.tgt")
+  # Generating (normal) test data (easy and hard).
+  generate_original_data_files(examples[2][0], examples[2][1], 
+                               data_path + "test_easy.src",
+                               data_path + "test_easy.tgt")
+  generate_original_data_files(examples[4][0], examples[4][1], 
+                               data_path + "test_hard.src",
+                               data_path + "test_hard.tgt")
+  # Generating iterative decoding training data.
+  generate_it_dec_data_files(examples[1][0], examples[1][1], 
+                               data_path + "it_dec_train.src",
+                               data_path + "it_dec_train.tgt")
+  # Generating iterative decoding val and test data (easy).
+  ratio = 0.1
```

Comment:
What is this "ratio" used for? is it to split train/test?

Reply:
This is the ratio of test samples that will be used for "validation", i.e., to calculate step-by-step accuracy.